File: go/tournament/tournament.go
package tournament

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"sort"
	"strings"
)

type Teams []*Team

func (t Teams) Len() int {
	return len(t)
}

func (t Teams) Swap(i, j int) {
	t[i], t[j] = t[j], t[i]
}

func (t Teams) Less(i, j int) bool {
	if t[i].Points() == t[j].Points() {
		return t[i].Name < t[j].Name
	}
	return t[i].Points() > t[j].Points()
}

func (t Teams) String() string {
	var s string
	for _, v := range t {
		s += fmt.Sprint(v)
	}
	return s
}

type Header struct{}

func (h *Header) String() string {
	return fmt.Sprintf("%-30s | %2s | %2s | %2s | %2s | %2s\n", "Team", "MP", "W", "D", "L", "P")
}

type Team struct {
	Name                     string
	Played, Won, Drawn, Lost int
}

func (t *Team) Play() {
	t.Played++
}

func (t *Team) Win() {
	t.Won++
}

func (t *Team) Lose() {
	t.Lost++
}

func (t *Team) Draw() {
	t.Drawn++
}

func (t *Team) Points() int {
	return t.Won*3 + t.Drawn*1
}

func (t *Team) String() string {
	return fmt.Sprintf("%-30s | %2d | %2d | %2d | %2d | %2d\n", t.Name, t.Played, t.Won, t.Drawn, t.Lost, t.Points())
}

var teams = make(map[string]*Team)

func clear() {
	teams = make(map[string]*Team)
}

func Tally(in io.Reader, out io.Writer) error {
	defer clear()

	reader := bufio.NewReader(in)
	for {
		line, _, err := reader.ReadLine()
		if err != nil {
			break
		}
		if len(line) == 0 {
			continue
		}

		parse := strings.Split(string(line), ";")
		if len(parse) != 3 {
			continue
		}

		apply0, apply1, err := GetAcrion(parse[2])
		if err != nil {
			continue
		}
		PlayedTeam(parse[0], apply0)
		PlayedTeam(parse[1], apply1)
	}

	if (len(teams)) == 0 {
		return errors.New("empty")
	}

	fmt.Fprint(out, new(Header))

	results := Teams{}
	for _, v := range teams {
		results = append(results, v)
	}
	sort.Sort(results)
	fmt.Fprint(out, results)

	return nil
}

func GetAcrion(result string) (apply0, apply1 func(*Team), err error) {
	switch result {
	case "win":
		apply0 = func(t *Team) { t.Win() }
		apply1 = func(t *Team) { t.Lose() }
	case "loss":
		apply0 = func(t *Team) { t.Lose() }
		apply1 = func(t *Team) { t.Win() }
	case "draw":
		apply0 = func(t *Team) { t.Draw() }
		apply1 = func(t *Team) { t.Draw() }
	default:
		err = errors.New(result)
	}
	return
}

func PlayedTeam(name string, fn func(*Team)) {
	team, ok := teams[name]
	if ok == true {
		team.Play()
		fn(team)
	} else {
		newTeam := &Team{Name: name}
		teams[name] = newTeam
		newTeam.Play()
		fn(newTeam)
	}
}

File: elixir/tournament/tournament.exs
defmodule Tournament do
  @doc """
  Given `input` lines representing two teams and whether the first of them won,
  lost, or reached a draw, separated by semicolons, calculate the statistics
  for each team's number of games played, won, drawn, lost, and total points
  for the season, and return a nicely-formatted string table.

  A win earns a team 3 points, a draw earns 1 point, and a loss earns nothing.

  Order the outcome by most total points for the season, and settle ties by
  listing the teams in alphabetical order.
  """
  @spec tally(input :: list(String.t())) :: String.t()
  def tally(input) do
    input
    |> Enum.map(&parse/1)
    |> Enum.reduce(%{}, &register/2) |> Map.values
    |> Enum.sort(&order/2)
    |> output
  end

  defp parse(input) do
    Regex.named_captures(~r/(?<teama>[\p{L}|\s]+);(?<teamb>[\p{L}|\s]+);(?<result>(win|loss|draw))$/, input)
  end

  defp register(nil, acc), do: acc
  defp register(%{"teama" => teama, "teamb" => teamb, "result" => result}, acc) do
    acc
    |> Map.put_new(teama, new_record(teama))
    |> Map.get_and_update(teama, fn r -> {r, update_team_result(r, result)} end) |> elem(1)
    |> Map.put_new(teamb, new_record(teamb))
    |> Map.get_and_update(teamb, fn r -> {r, update_opponent_result(r, result)} end) |> elem(1)
  end

  defp new_record(name) do
    %{name: name, played: 0, won: 0, drawn: 0, lost: 0, points: 0}
  end

  defp update_team_result(record, "win"),  do: record |> update_point(3) |> update_count(:won)
  defp update_team_result(record, "draw"), do: record |> update_point(1) |> update_count(:drawn)
  defp update_team_result(record, "loss"), do: record |> update_point(0) |> update_count(:lost)
  defp update_opponent_result(record, "win"),  do: update_team_result(record, "loss")
  defp update_opponent_result(record, "draw"), do: update_team_result(record, "draw")
  defp update_opponent_result(record, "loss"), do: update_team_result(record, "win")

  defp update_point(record, amount) do
    record
    |> update_in([:points], &(&1 + amount))
  end

  defp update_count(record, result) do
    record
    |> update_in([:played], &(&1 + 1))
    |> update_in([result], &(&1 + 1))
  end

  defp order(%{points: points_a}, %{points: points_b}) when points_a != points_b do
    points_a > points_b
  end

  defp order(%{name: name_a}, %{name: name_b}) do
    name_a < name_b
  end

  defp output(records) do
    header = %{name: "Team", played: "MP", won: "W", drawn: "D", lost: "L", points: "P"}
    [header | records] |> Enum.map(&output_record/1) |> Enum.join("\n")
  end

  defp output_record(%{name: team, played: mp, won: w, drawn: d, lost: l, points: p}) do
    [
      String.pad_trailing(team, 30),
      String.pad_leading(to_string(mp), 2),
      String.pad_leading(to_string(w), 2),
      String.pad_leading(to_string(d), 2),
      String.pad_leading(to_string(l), 2),
      String.pad_leading(to_string(p), 2)
    ]
    |> Enum.join(" | ")
  end
end
